### PR TITLE
[3.5] bpo-39503: CVE-2020-8492: Fix AbstractBasicAuthHandler (GH-18284)

### PR DESCRIPTION
**Pull request by vstinner**
The AbstractBasicAuthHandler class of the urllib.request module uses
an inefficient regular expression which can be exploited by an
attacker to cause a denial of service. Fix the regex to prevent the
catastrophic backtracking. Vulnerability reported by Ben Caller
and Matt Schwager.

AbstractBasicAuthHandler of urllib.request now parses all
WWW-Authenticate HTTP headers and accepts multiple challenges per
header: use the realm of the first Basic challenge.

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>
(cherry picked from commit 0b297d4ff1c0e4480ad33acae793fbaf4bf015b4)

https://bugs.python.org/issue39503

### CODE REVIEW

**Comment by vstinner** [comment]
For the backport, I replaced f-string with str.format() in tests.

**Comment by vstinner** [comment]
PR rebased on top of commit ed07522a5faa3101f68be8e4b8369310f60860f8 (remove "Codecov patch" job from the CI).

**Comment by vstinner** [comment]
cc @larryhastings: the CI is now green.

**Comment by vstinner** [comment]
@larryhastings: This change is backward incompatible. If a HTTP header contains two Basic challenges, now the first one is used, whereas previously the last one was chosen. IMO the new behavior is more correct than the old one, but I prefer to warn you ;-)

**Comment by vstinner** [comment]
Ping @larryhastings: Would you mind to merge this backport to 3.5 of a security fix?

File: Lib/urllib/request.py [hunk 1]
```diff
@@ -885,8 +885,15 @@ class AbstractBasicAuthHandler:
 
     # allow for double- and single-quoted realm values
     # (single quotes are a violation of the RFC, but appear in the wild)
-    rx = re.compile('(?:.*,)*[ \t]*([^ \t]+)[ \t]+'
-                    'realm=(["\']?)([^"\']*)\\2', re.I)
+    rx = re.compile('(?:^|,)'   # start of the string or ','
+                    '[ \t]*'    # optional whitespaces
+                    '([^ \t]+)' # scheme like "Basic"
+                    '[ \t]+'    # mandatory whitespaces
+                    # realm=xxx
```
Comment (larryhastings) [inline Lib/urllib/request.py:892]: Could you remove this, please?

**Review by larryhastings — DISMISSED** [review]

File: Lib/urllib/request.py (hunk 1 above)
Comment (vstinner) [reply Lib/urllib/request.py:892]: I'm sorry, I don't understand your request.

This PR is a straightforward backport of https://github.com/python/cpython/commit/0b297d4ff1c0e4480ad33acae793fbaf4bf015b4 The code in master also contains the comments like "realm=xxx".

The following comment shows which formats are accepted for the realm field:

```
# realm=xxx
# realm='xxx'
# realm="xxx"
```

See unit tests for more concrete examples.

If you want to remove/rephrase the comments, it should be done first in the master branch.

Comment (larryhastings) [reply Lib/urllib/request.py:892]: Oh, okay.  I thought this was debug test code that got left in, like leaving a scalpel in the patient. ;-)

**Pull request merged**